### PR TITLE
Add users list for claims school page

### PR DESCRIPTION
**Pull request by gms-gs**
## Context

This change adds the Users flow to the claims school page

- Sign in as a lead mentor claims user
- Go to `http://claims.localhost:3000/schools/{id}/users` to see a list of users
- Add a user
- Go to the users index page
- Find the added user
- Click on the users name 

## Link to Trello card

https://trello.com/c/Lxs1LngQ/97-create-the-school-users-page

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated or [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

## Screenshots
Screenshot 2024-01-23 at 07 14 59
Screenshot 2024-01-23 at 07 15 20
Screenshot 2024-01-23 at 07 15 25
Screenshot 2024-01-23 at 07 15 29
Screenshot 2024-01-23 at 07 15 35

### CODE REVIEW

File: app/controllers/claims/schools/users_controller.rb [hunk 1]
```diff
@@ -1,3 +1,40 @@
 class Claims::Schools::UsersController < ApplicationController
   include Claims::BelongsToSchool
+
+  def index
+    @users = @school.users
```
Comment (JamieCleare2525) [inline app/controllers/claims/schools/users_controller.rb:5]:
```suggestion
    @users = @school.users.claims
```

File: app/controllers/claims/schools/users_controller.rb [hunk 2]
```diff
@@ -1,3 +1,40 @@
...
+  end
+
+  def show
+    @user = @school.users.find(params.require(:id))
```
Comment (JamieCleare2525) [inline app/controllers/claims/schools/users_controller.rb:9]:
```suggestion
    @user = @school.users.claims.find(params.require(:id))
```

File: app/controllers/claims/schools/users_controller.rb (hunk 1 above)
Comment (JamieCleare2525) [reply app/controllers/claims/schools/users_controller.rb:5]: Might be worth doing something like:
```
def users
  @users ||= @school.users.claims
end

File: spec/system/claims/invite_a_user_to_a_school_spec.rb [hunk 3]
```diff
@@ -41,15 +47,15 @@
 
   scenario "I add a user to different schools" do
```
Comment (JamieCleare2525) [inline spec/system/claims/invite_a_user_to_a_school_spec.rb:48]: Surely, a general user should only be able to add users to their own schools.

**Comment by JamieCleare2525** [comment]
> > Can we add Rspec tests for `user_views_a_list_of_users_in_their_school` and `user_views_another_users_details`?
> 
> Does this test file not test what we need already? `spec/system/claims/invite_a_user_to_a_school_spec.rb` @JamieCleare2525
> 
> It tests details appear ect

It doesn't specifically test that only users in for a specific organisation can be seen by a user in that organisation.
It also doesn't test the links to the show pages from the index page.

I would also expect some form of security to be introduced here. As I would expect general users should only see schools that they are assigned to.

File: app/controllers/claims/schools/users_controller.rb (hunk 1 above)
Comment (Nitemaeric) [reply app/controllers/claims/schools/users_controller.rb:5]: To add additional context here. The `Claims::School` redefines the `has_many :users` scope to include the `.claims` scope already.

This means, any `Claims::School#users` calls will already scope the users by the claims service.

File: app/views/claims/schools/users/check.html.erb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        <%= f.hidden_field :first_name, value: @user_form.first_name %>
+        <%= f.hidden_field :last_name, value: @user_form.last_name %>
+        <%= f.hidden_field :email, value: @user_form.email %>
+
+        <label class="govuk-label govuk-label--l">
+          <span class="govuk-caption-l"><%= t(".caption", school_name: @school.name) %></span>
+          <%= t(".title") %>
+        </label>
+
+        <dl class="govuk-summary-list">
```
Comment (elceebee) [inline app/views/claims/schools/users/check.html.erb:24]: could we place use the govuk_summary_list component here? See [here](https://github.com/DFE-Digital/itt-mentor-services/blob/ec5fcd1c683c9e5373473a93718bb3204e10b851/app/views/placements/support/organisations/users/_check_form.html.erb
)
Better yet, move that component to a shared space and use it directly so we don't have so much repetition.

File: app/views/claims/schools/users/index.html.erb [hunk 5]
```diff
@@ -4,4 +4,30 @@
 
 <div class="govuk-width-container">
   <h1 class="govuk-heading-l"><%= t(".heading") %></h1>
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <%= govuk_button_to(t(".add_user"), new_claims_school_user_path, method: :get) %>
+
+      <% if @users.any? %>
+        <table class="govuk-table">
```
Comment (elceebee) [inline app/views/claims/schools/users/index.html.erb:12]: Could we use the govuk_table component? See [here](https://github.com/DFE-Digital/itt-mentor-services/blob/d8b74634014de56998d6cf585466844257435610/app/views/placements/support/schools/users/index.html.erb)
Again, it's now something that is used in many places, so good time to make it a shared component. And in doing so, you'll fix the empty state problem everywhere at once.

File: app/views/claims/schools/users/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  <%= govuk_back_link(href: claims_school_users_path(@school)) %>
+<% end %>
+
+<div class="govuk-width-container">
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <span class="govuk-caption-l"><%= @school.name %></span>
+      <h2 class="govuk-heading-l"><%= @user.full_name %></h2>
+
+      <dl class="govuk-summary-list">
```
Comment (elceebee) [inline app/views/claims/schools/users/show.html.erb:11]: Again, let's use the govuk_summary list component. See [here](https://github.com/DFE-Digital/itt-mentor-services/blob/7758930db2dbabbf75e05d07541e45c75dd78999/app/views/placements/support/organisations/users/_details.html.erb)

And better yet, move that component to a the shared folder and just use it directly so we don't have so much repetition.

**Review by elceebee — COMMENTED** [review]
I agree with Jamie that we need to check about the requirements -- unless it's a support user, they should only be able to add users to their own school, right?

And would be really nice to reduce some of the repetition through refactoring of reused code. If not, at least use the govuk components to make the code more concise and consistent.

File: app/controllers/claims/schools/users_controller.rb (hunk 1 above)
Comment (CatalinVoineag) [inline app/controllers/claims/schools/users_controller.rb:5]: Sorry to open this again, I think we have scoping issue. If you know the ID of a school of a different user you can go to that user's school. I'm not saying we should fix this in this PR as I expect we have similar scoping issues in other places. But to me this feels like where policies come in.

I would expect an index policy here that checks the school_id in the param against the current user schools. Something like:
```
School policy 

def index
  current_user.schools.pluck(:id).inlcude?(params.require(:school_id))
end
```


https://github.com/DFE-Digital/itt-mentor-services/assets/11318084/a0f5a973-363a-43a4-9d09-43e7f2cb9765

Don't want to block this PR, but we should probably address this either here or in a different PR

Comment (CatalinVoineag) [reply app/controllers/claims/schools/users_controller.rb:5]: Thinking about it more, we can also do this scoping in the controller, we just need to handle the user not being allowed to view another school gracefully. Don't mind how we solve this.

Comment (Nitemaeric) [reply app/controllers/claims/schools/users_controller.rb:5]: @CatalinVoineag I like the idea of adding Pundit in, but maybe only when we have resource-based permissions in place. Right now, any user belonging to a school can edit any other user, claim, etc. For now, we can likely get by by scoping down the lookup of school.

@gms-gs 
We could update the `Claims::BelongsToSchool` to scope down the top-level scope to `current_user.schools.find_by(params.require(:id))`. The challenge here is to now make sure it's using the right namespace.

```diff
class Claims::User < User
  default_scope { claims }

+ has_many :schools,
+          through: :memberships,
+          source: :organisation,
+          source_type: "School"
end
```

Adding this in should make sure `current_user.schools` uses the `Claims::School` model as long as `current_user` is of class `Claims::User`. As far as I can tell, the codebase currently does this namespacing.

Comment (ollietreend) [reply app/controllers/claims/schools/users_controller.rb:5]: Please just be aware that `Claims::BelongsToSchool` is also used by the support controllers.

As a rule, support users will need access to _all_ schools. But it seems wasteful & like unnecessary overhead to set up membership records linked every support user with every school.

Instead, I think you'll probably want a `Claims::SupportUser` which redefines the association to `has_many :schools, -> { claims }` so every school is available on `current_user`.

File: app/controllers/concerns/claims/belongs_to_school.rb [hunk 7]
```diff
@@ -5,7 +5,16 @@ module Claims::BelongsToSchool
     before_action :set_school
 
     def set_school
-      @school = Claims::School.find(params.require(:school_id))
+      @school = scoped_schools.find(params.require(:school_id))
+    rescue ActiveRecord::RecordNotFound
+      flash[:alert] = "School not found"
+      redirect_to claims_schools_path
```
Comment (Nitemaeric) [inline app/controllers/concerns/claims/belongs_to_school.rb:11]: Didn't we decide to let this go to a 404 error page? 🤔

File: config/environments/development.rb [hunk 8]
```diff
@@ -12,7 +12,7 @@
   config.eager_load = false
 
   # Show full error reports.
-  config.consider_all_requests_local = true
+  config.consider_all_requests_local = false
```
Comment (Nitemaeric) [inline config/environments/development.rb:15]: 🙊

File: config/locales/en/claims/schools/users.yml [hunk 9]
```diff
@@ -4,3 +4,39 @@ en:
       users:
         index:
           heading: Users
+          add_user: Add user
+          no_users: There are no users for %{school_name}.
+          attributes:
+            users:
+              name:  Name
+              email: Email
```
Comment (Nitemaeric) [inline config/locales/en/claims/schools/users.yml:12]: Instead of adding these attributes localisation keys in. Please use `User.human_attribute_name(attribute_name)`. e.g. `User.human_attribute_name(:first_name)` or `User.human_attribute_name(:email)`.

This is create consistency in how we present the attributes names to users.

**Review by CatalinVoineag — APPROVED** [review]
Once Dan's comments are addressed I think this is ready to go 🚀

File: config/environments/development.rb (hunk 8 above)
Comment (elceebee) [reply config/environments/development.rb:15]: Why are we changing this? Surely we want this to be true for development so we can see more detailed error logs for debugging? 
https://guides.rubyonrails.org/configuring.html#config-consider-all-requests-local

File: config/locales/en/activerecord.yml [hunk 10]
```diff
@@ -18,3 +19,4 @@ en:
               blank: Enter an email address
               taken: This email address is already in use. Try another email address
               invalid: Enter an email address in the correct format, like name@example.com
+              invalid_support_email: Enter a Department for Education email address in the correct format, like name@education.gov.uk
```
Comment (Nitemaeric) [inline config/locales/en/activerecord.yml:22]: i think this is out of scope.

**Review by Nitemaeric — APPROVED** [review]
1 thing that slipped in, all else looks good!
